File: solutions/200_NumberofIslands.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        result = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    self.dfs(grid,i,j)
                    result += 1
        return result
    
    def dfs(self,grid,i,j):
        if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]) or grid[i][j] == '0':
            return
        
        grid[i][j] = '0'
        self.dfs(grid, i + 1, j)
        self.dfs(grid, i-1, j)
        self.dfs(grid, i, j+1)
        self.dfs(grid, i , j-1)

File: solutions/200_NumberofIslands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    let count = 0;
    function dfs(grid, i, j){
        if(i<0 || i >= grid.length || j < 0 || j >=grid[0].length || grid[i][j] != '1'){
            return;
        }
        grid[i][j] = '#' // mark as visited can be marked as 0 
        dfs(grid,i+1,j);
        dfs(grid,i-1,j);
        dfs(grid,i,j+1);
        dfs(grid,i,j-1);
    }
    for(let i = 0;i<grid.length;i++){
        for(let j =0;j<grid[0].length;j++){
            if(grid[i][j]=='1'){
                dfs(grid,i,j)
                count = count + 1
            }
        }
    }
    return count
};